### ISSUE TITLE
JSON translations don't work in SSR Spartacus 4.x

### ISSUE THREAD

**Issue by fgrecchi-aswatson** [SAP/spartacus#14372]
Hello,

I am opening a bug after creating a topic [here](https://stackoverflow.com/questions/69977308/backend-translations-dont-work-in-ssr-spartacus).

I have implemented the Lazy Loading of the translations as in this [guide](https://sap.github.io/spartacus-docs/i18n/#lazy-loading). If I run the application in SSR mode it always falls back in CSR, even if I increase the timeout to over 1 minute.
If I remove the backend configuration and restore the static files, the SSR works fine.
This issue occurs on both version 4.1 and 4.2 of Spartacus.

Steps to reproduce:
```
ng new electronics-spa --style=scss --package-manager yarn
cd electronics-spa/
ng add @spartacus/schematics@latest --baseUrl https://spartacus-demo.eastus.cloudapp.azure.com:8443 --baseSite=electronics-spa --currency=USD --language=en --ssr --skip-confirmation
cp ./node_modules/@spartacus/assets/i18n-assets ./src/assets -r
```

(I am using of course mine SAP Cloud baseUrl)

Change the file `src/app/spartacus/spartacus-configuration.module.ts` in this way (as per the documentation):

```
providers: [
  provideConfig({
    i18n: {
      backend: {
        loadPath: 'assets/i18n-assets/{{lng}}/{{ns}}.json',
      },
      chunks: translationChunksConfig,
    },
  }),
];
```

Then finally run
```
yarn build:ssr
yarn serve:ssr
```

And when you open the application in the browser it will fall back in the CSR mode without any error on server side.
If you revert the spartacus-configuration.module.ts and re-build and re-run the application, SSR works fine.

I have created a simple repo that reproduces the error: https://github.com/fgrecchi-aswatson/spartacus-ssr-issue

Thanks
Filippo Grecchi

**Comment by Xymmer** [comment]
seeking feedback from dev time, will come back soon

**Comment by Platonn** [comment]
No simple answer for this ticket from top of my head. I’ll need to dive and debug.
Many thanks @fgrecchi-aswatson for providing a REPO for reproducing!

**Comment by fgrecchi-aswatson** [comment]
Hello,

a coworker of mine find out this possible solution related to i18next configuration

In `projects/core/src/i18n/i18next/i18next-init.ts` we added `reloadInterval: 0` to the configuration object passed to `i18next.init()`.

So here's the code:

```
...
          if (config.i18n.backend) {
            i18next.use(i18nextHttpBackend);
            const loadPath = getLoadPath(
              config.i18n.backend.loadPath,
              serverRequestOrigin
            );
            const backend = {
              loadPath,
              ajax: i18nextGetHttpClient(httpClient),
              reloadInterval: 0,
            };
            i18nextConfig = { ...i18nextConfig, backend };
          }
...
```

I hope this might help.
Filippo

**Comment by Platonn** [comment]
[work in progress]
I'm documenting here my process of debugging for future reference:
1. clone reproduction repo and run `yarn`
1. pass additional configuration in the 2nd param of `get()` in `server.ts`:
   ```ts
   const ngExpressEngine = NgExpressEngineDecorator.get(engine, {
      debug: true,
      timeout: 20000,
      maxRenderTime: 30000,
   });
   ```
1. configure `i18n.debug: true` in Spartacus `SpartacusConfigurationModule`: 
   ![image](https://user-images.githubusercontent.com/4001059/142891412-c563c61c-13a1-491f-8532-e01d59677d98.png)

1. run `yarn build:ssr`
1. run `yarn serve:ssr`
1. create a launch.json file (if needed)
   ![image](https://user-images.githubusercontent.com/4001059/142893618-b505241a-80f1-4f4b-8aac-3e08e33f5f18.png)
   Add a new debug configuration option in dropdown:
   ![image](https://user-images.githubusercontent.com/4001059/142893853-aa78522a-221c-4508-8411-96966898045d.png)
   ![image](https://user-images.githubusercontent.com/4001059/142893959-6fbc5b11-10fe-42dd-a88e-4a189fc76565.png)
   Choose "attach to process"
   ![image](https://user-images.githubusercontent.com/4001059/142894167-8842a337-4b35-4ea5-9f83-ab819425b871.png)

**Comment by Platonn** [comment]
I confirm it's a bug. The JSON translations work in CSR, but don't work OOTB in SSR:
- for schematics-installed Spartacus 4.2.0
- ~for schematics-installed Spartacus 3.4.5~
- for develop 4.2 (our repo)


To reproduce in our repo:
`yarn build:assets`
`cp -r dist/assets/i18n-assets projects/storefrontapp/src/assets`

config:
```diff
i18n: {
-    resources: translations,
+    backend: {
+          loadPath: 'assets/i18n-assets/{{lng}}/{{ns}}.json',
+    },
}
```

**Comment by Platonn** [comment]
Hi @fgrecchi-aswatson. I confirm, the fix is to set `reloadInterval: false` in the backend options of i18next
Unfortunately until https://github.com/SAP/spartacus/issues/11675 is implemented, it's cumbersome for customers to apply quick fix themselves. Currently, to change any logic in the function `18nextInit()` customers need to dismantle deeply the modules tree up to the `I18nModule` and then provide a custom factory function `i18nextInit()`

Detailed results of investigation:
In 4.0, when switching from the deprecated 3rd party dependency `i18next-xhr-backend` to its new counterpart  `i18next-http-backend` we didn't notice that there are slight differences between those packages:
- the new package has a bit different API for passing the custom http client - which was already addressed in `develop` branch by this PR https://github.com/SAP/spartacus/pull/14453/files#diff-b9c2f7d32db998292ce5b32da446437ad64755d84817c3261fdbc89f6438856aL99-R113 
- the new package invokes the function `setInterval()` which makes the SSR render to hang. This issue will be addressed by the PR to `develop` https://github.com/SAP/spartacus/pull/15297

Since `develop` branch now is aimed to be released only as 5.0, we should backport both PRs https://github.com/SAP/spartacus/pull/14453/ and https://github.com/SAP/spartacus/pull/15297 to `release/4.3.x` and then release a new `4.3.x` patch.

This is a priority task now.

**Comment by Platonn** [comment]
For reference:
The fixes have been already released in v4.3.1. For more, see the [commits history of release/4.3.x. branch](https://github.com/SAP/spartacus/commits/release/4.3.x).
